wagtail/core/blocks/stream_block.py

```diff
@@ -424,6 +419,33 @@ def _prefetch_blocks(self, type_name, child_block):
             block_id = self.stream_data[i].get('id')
             self._bound_blocks[i] = StreamValue.StreamChild(child_block, value, id=block_id)
 
+    def get_prep_value(self):
+        prep_value = []
+
+        for i, stream_data_item in enumerate(self.stream_data):
+            if self.is_lazy and i not in self._bound_blocks:
+                # This child has not been accessed as a bound block, so its raw JSONish
+                # value (stream_data_item here) is still valid
+
+                # assign a new ID if it didn't have one already
+                if not stream_data_item.get('id'):
+                    stream_data_item['id'] = str(uuid.uuid4())
+
+                prep_value_item = stream_data_item
+
+            else:
+                # convert the bound block back into JSONish data
+                child = self[i]
+                prep_value_item = {
+                    'type': child.block.name,
+                    'value': child.block.get_prep_value(child.value),
+                    # assign a new ID on save if it didn't have one already
+                    'id': child.id or str(uuid.uuid4()),
```

Comment:
It'd be nice if the `str(uuid.uuid4())` didn't need to be repeated. We could also have a clearer comment about why creation of this ID is desirable. What about something like:

```py
prep_value = []

for i, stream_data_item in enumerate(self.stream_data):
    if self.is_lazy and i not in self._bound_blocks:
        # This child has not been accessed as a bound block, so its raw JSONish
        # value (stream_data_item here) is still valid
        prep_value_item = stream_data_item
    else:
        child = self[i]
        prep_value_item = {
            'type': child.block.name,
            'value': child.block.get_prep_value(child.value),
            'id': child.id,
        }

    # As this method is preparing this value to be saved to the database,
    # this is an appropriate place to ensure that each block has a unique id.
    if not prep_value_item.get('id'):
        prep_value_item['id'] = str(uuid.uuid4())

    prep_value.append(prep_value_item)

...
```